Source/playerdat.cpp

```diff
@@ -6,69 +6,287 @@
+		12042092,
+		15571031,
+		20066900,
+		25774405,
+		32994399,
+		42095202,
+		53525811,
+		67831218,
+		85670061,
+		107834823,
+		135274799,
+		169122009,
+		210720231,
+		261657253,
+		323800420,
+		399335440,
+		490808349,
+		601170414,
+		733825617,
+		892680222,
+		1082908612,
+		1310707109,
+		1583495809
+	};
+
+public:
+	uint8_t getMaxLevel() const
+	{
+		if (levelThresholds.empty())
+			return 0;
+		return static_cast<uint8_t>(std::min<size_t>(levelThresholds.size() - 1, std::numeric_limits<uint8_t>::max()));
+	}
+
+	DVL_REINITIALIZES void clear()
+	{
+		levelThresholds.clear();
+	}
+
+	[[nodiscard]] uint32_t getThresholdForLevel(unsigned level) const
+	{
+		return levelThresholds[std::min<size_t>(level, getMaxLevel())];
+	}
+
+	void setThresholdForLevel(unsigned level, uint32_t experience)
+	{
+		if (level >= levelThresholds.size()) {
+			// To avoid ValidatePlayer() resetting players to 0 experience we need to use the maximum possible value here
+			// As long as the file has no gaps it'll get initialised properly.
+			levelThresholds.resize(static_cast<size_t>(level) + 1, std::numeric_limits<uint32_t>::max());
+		}
+		levelThresholds[level] = experience;
+	}
+} ExperienceData;
+
+struct ExperienceColumnDefinition {
+	enum class ColumnType {
+		Level,
+		Experience,
+		LAST = Experience
+	} type;
+
+	enum class Error {
+		UnknownColumn
+	};
+
+	// The number of fields between this column and the last one identified as important (or from start of the record if this is the first column we care about)
+	unsigned skipLength;
+
+	static tl::expected<ColumnType, Error> mapNameToType(std::string_view name)
+	{
+		if (name == "Level") {
+			return ColumnType::Level;
+		}
+		if (name == "Experience") {
+			return ColumnType::Experience;
+		}
+		return tl::unexpected { Error::UnknownColumn };
+	}
+
+	ExperienceColumnDefinition() = delete;
+
+	ExperienceColumnDefinition(const ColumnType &type)
+	    : type(type)
+	    , skipLength(0)
+	{
+	}
+
+	ExperienceColumnDefinition(const ColumnType &type, unsigned skipLength)
+	    : type(type)
+	    , skipLength(skipLength)
+	{
+	}
+
+	bool operator==(const ExperienceColumnDefinition &other) const
+	{
+		return type == other.type && skipLength == other.skipLength;
+	}
 };
 
+void ReloadExperienceData()
+{
+	auto dataFileResult = DataFile::load("txtdata\\Experience.tsv");
+	if (!dataFileResult.has_value()) {
+		// Can't load the data file, so reset to defaults.
+		// TODO: Log error somewhere the player can see it, they probably want to know that we couldn't load the file
+		ExperienceData = {};
+		return;
+	}
+	const DataFile &dataFile = dataFileResult.value();
+
+	constexpr unsigned ExpectedColumnCount = enum_size<ExperienceColumnDefinition::ColumnType>::value;
+
+	StaticVector<ExperienceColumnDefinition, ExpectedColumnCount> columns;
+	std::bitset<ExpectedColumnCount> seenColumns;
+
+	unsigned currentColumn = 0;
+	unsigned lastColumn = 0;
+	GetFieldResult result { dataFile.begin() };
+	do {

```

Comment:
Looks like this loop can be rewritten with the new iterators?

Reply:
For example, parsing a row would go something like this:

```c++
auto it = records.begin();
const auto end = records.end();

if (it == end) { /* not enough columns */ }

// parse key

++it;
if (it == end) { /* not enough columns */ }

// parse value
```

Reply:
That would require scanning the first record, parsing over the fields in that record (scanning it again), then advancing past the first record (scanning it a third time) when consuming the values. Didn't seem worth it unless we want to make the rest of the code more fragile.

Reply:
We can have a `skipHeader()` method that returns a `{++begin, end}` range.

Alternatively, this can also be done manually (once we add a pair of iterators constructor for range):

```c++
auto recordsBegin = records.begin();
const auto recordsEnd = records.end();
if (recordsBegin == recordsEnd) { /* empty file */ };

// handle the header at `*recordsBegin`

for (const auto &record : RecordsRange{++recordsBegin, recordsEnd}) {
 // non-header rows
}
```

Reply:
This code will be about the same length as the `GetNext*` version with the major advantage being that iterator semantics are well-known.

Reply:
In the current iterator code the RecordsRange::Iterator scans to the end of the record and provides a string_view that the FieldsInRecordRange::Iterator scans again to extract individual fields. The obvious alternative would be to make the RecordsRange::Iterator chunk the record into fields and hold onto that state, then the field-based iterator would be derived from whatever internal container is used, I wasn't sure if having a heavy iterator like that would be suitable.

If you have another approach in mind I'd be interested in seeing how it can be done in a single pass (can the field iterator mutate state in the record iterator so the record iterator doesn't actually need to find the end of the record in advance?)

Reply:
In the past I've had success with the following approach for parsers:

1. The `RecordsRange::Iterator` holds on to a single `FieldsInRecordRange::Iterator` (let's call it `field_iterator_`).
2. The `RecordsRange::begin()` method returns a mutable reference to `field_iterator_`. So as we advance the `field_iterator_`.
3. To increment a `RecordsRange::Iterator`, seek to the next EOL from the end of the string view of the current `field_iterator_`.

Reply:
Another thing I've done in the past is keep a single position in the Record iterator and have the Field iterator keep a pointer to that position. This allows begin() and end() to return by-value (more intuitive) but makes the field iterator an Input iterator (single-pass) rather than a Forward iterator (multi-pass) -- this may be better for this use case

Reply:
I don't mind merging this as-is and implementing the iterator myself btw if you prefer

Reply:
Sure, I'll probably not get around to redoing the iterator until next weekend. I've added a few more test cases to try get full coverage (or as close as I can without being able to trigger a read error) which should at least help validate the behaviour of the input iterator approach.